torch/csrc/api/include/torch/data/ctf/ctf_parser.h

```diff
@@ -0,0 +1,931 @@
+      CTFDataType data_type)
+      : filepath_(std::move(filepath)),
+        input_streams_info_(std::move(input_streams_info)),
+        data_type_(data_type){};
+
+  const std::vector<CTFInputStreamInformation>& get_input_streams_info() const {
+    return input_streams_info_;
+  }
+
+  const std::string& get_file_path() const {
+    return filepath_;
+  }
+  CTFDataType get_ctf_data_type() const {
+    return data_type_;
+  }
+
+ private:
+  std::string filepath_;
+  std::vector<CTFInputStreamInformation> input_streams_info_;
+  CTFDataType data_type_;
+};
+
+///
+/// Sequence ID type
+/// -1 is used to flag an uninitialized Sequence ID
+///
+typedef long int CTFSequenceID;
+
+#ifdef CTF_DEBUG
+///
+/// Maps Sequenced ID to index at vector<CTFSequenceData>
+///
+typedef std::map<size_t, size_t> CTFSequenceMap;
+#endif
+
+///
+/// Input Stream ID type
+/// All Input Streamsare stored on a vector<CTFINputStreamInformation>
+/// and CTFInputStreamID is the index of a particular stream
+///
+typedef size_t CTFInputStreamID;
+
+///
+/// Maps Input Stream names to a unique index
+///
+typedef std::unordered_map<std::string, CTFInputStreamID>
+    CTFInputStreamMapByName;
+
+///
+/// Used during sparse data parsing
+///
+const size_t CTFValueIndexUninitialized = SIZE_MAX;
+typedef size_t CTFValueIndex;
+
+///
+/// Sequence data type
+/// The global vector of sequences and the vector of samples will use it
+///
+struct CTFInpuStreamDataBase {
+  explicit CTFInpuStreamDataBase(size_t input_stream_id)
+      : input_stream_id(input_stream_id) {}
+  size_t input_stream_id;
+};
+typedef std::shared_ptr<CTFInpuStreamDataBase> CTFInpuStreamDataBasePtr;
+typedef std::vector<CTFInpuStreamDataBasePtr> CTFSequenceData;
+
+///
+/// Dense data
+///
+template <typename DataType>
+struct CTFDenseInputStreamData : CTFInpuStreamDataBase {
+  explicit CTFDenseInputStreamData(size_t input_stream_id, size_t capacity = 0)
+      : CTFInpuStreamDataBase(input_stream_id) {
+    if (capacity > 0) {
+      // TODO: On a per input stream storage, sample dimension is not useful
+      // data.reserve(capacity);
+    }
+  }
+
+  std::vector<DataType> data;
+};
+
+///
+/// Sparse data
+///
+template <typename DataType>
+struct CTFSparseInputStreamData : CTFInpuStreamDataBase {
+  explicit CTFSparseInputStreamData(
+      size_t input_stream_id,
+      size_t dimension = 0)
+      : CTFInpuStreamDataBase(input_stream_id) {
+    if (dimension > 0) {
+      /// TODO: Reserve something for sparse input? 1%? 0.05% of dimension?
+      // data.reserve(dimension);
+    }
+  }
+
+  std::vector<size_t> indices;
+  std::vector<DataType> data;
+};
+
+///
+/// CTFDataset centralizes all parsed data
+///
+template <typename DataType>
+struct CTFDataset {
+  explicit CTFDataset(
+      CTFDataType data_type,
+      const std::vector<CTFInputStreamInformation>& input_streams_info)
+      : data_type(data_type), input_streams_info(input_streams_info) {}
+
```

Comment:
Do you intend to move input_streams_info into the struct? Also, please remove the const keyword from the parameter input_streams_info to avoid falling back to copy constructor.

Reply:
As this is done just once, I guess I will just use copy constructor for input_streams_info and filepath